homeassistant/components/fan/reproduce_state.py

```diff
@@ -72,18 +81,46 @@ async def _async_reproduce_state(
     if state.state == STATE_ON:
         # The fan should be on
         if cur_state.state != STATE_ON:
-            # Turn on the fan at first
-            service_calls[SERVICE_TURN_ON] = service_data
-
-        for attr, service in ATTRIBUTES.items():
+            # Turn on the fan with all the speed and modes settings.
+            # The `turn_on` method will figure out in which mode to
+            # turn the fan on.
+            data = service_data.copy()
+            for attr in SPEED_AND_MODES_ATTRIBUTES:
+                if attr in state.attributes and state.attributes[attr] is not None:
```

Comment:
Given the None check is repeated a few times, I think it makes sense to just do this first at the top of the section before updating attributes, e.g.:
```
state_attributes = {k: v for k, v in state.attributes.items() if v is not None}
```